File: 070_climbing-stairs/solution_3.dart
class Solution {
  int climbStairs(int n) {
    if (n <= 1) return 1;

    int prev2 = 1; // Ways to reach step n-2
    int prev1 = 1; // Ways to reach step n-1
    int current = 0;

    for (int i = 2; i <= n; i++) {
      current = prev1 + prev2;
      prev2 = prev1;
      prev1 = current;
    }

    return current;
  }
}

// Function to run test cases
void runTests() {
  List<int> testCases = [2, 3, 4, 5, 6];
  List<int> expected = [2, 3, 5, 8, 13];
  Solution sol = Solution();
  
  for (int i = 0; i < testCases.length; i++) {
    int result = sol.climbStairs(testCases[i]);
    print(
        'Test Case ${i + 1}: Input: ${testCases[i]} | Expected: ${expected[i]} | Got: $result');
  }
}

void main() {
  runTests();
}

File: 292_nim-game/solution_3.py
class Solution:
    def canWinNim(self, n: int) -> bool:
        if n <= 3:
            return True
        dp = [False] * (n + 1)
        
        dp[1], dp[2], dp[3] = True, True, True
        
        for i in range(4, n + 1):
            dp[i] = not dp[i - 1] or not dp[i - 2] or not dp[i - 3]
        
        return dp[n]

# Test cases for Python solution
s = Solution()
print(s.canWinNim(4))  # False
print(s.canWinNim(1))  # True
print(s.canWinNim(2))  # True
print(s.canWinNim(5))  # True
